### ISSUE TITLE
Please reconsider your stance on the F-Droid version's Internet permission

### ISSUE THREAD

**Issue by puelp** [zyrouge/symphony#119]
Symphony looks to be a beautiful, modern player and a top contender for the best FOSS Android music player, but the "full network access" is a huge dealbreaker for privacy-conscious people such as myself.

I see [here](https://github.com/zyrouge/symphony/issues/84) that you said that you won't fix the issue, but I urge you to reconsider.

It is unfortunate to see your app being penalized with a glaring "Anti-features: This app promotes or depends entirely on a non-free network service" warning displayed on the app's F-Droid page, when the F-Droid version doesn't even need a self-updater, so the app is essentially being punished for no actual reason.

To my understanding, the F-Droid build system works independently and is its own self-contained thing, and as a result, many app developers are OK with the F-Droid versions of their apps having different signatures to the ones that they release on the Play Store or self-release on GitHub. I don't imagine users who installed the F-Droid version suddenly wanting to switch to the Play Store/GitHub version, so the incompatibility in that regard would not be that much of a problem, in my opinion.

I see that the solution provided was "disable the update checker", but if unchecking the 'Check for updates' toggle was all it took for the user to trust the app not to do anything potentially harmful with the Internet permission, then at that point the user may as well be using a closed-source music player, right? It's a matter of absolute principle, you see...your mind wouldn't be at complete ease unless the app was 100% FOSS and 100% offline. Symphony satisfies the former condition, and I really wish it satisfied the latter as well.

Thank you for reading, and I look forward to the day that Symphony can be the top FOSS, 100%-fully-offline music player on F-Droid.

**Comment by SunBreaker1** [comment]
Use Karma FW from F-Droid to block internet connections.

This should your very specific problem with the app

**Comment by puelp** [comment]
> Use Karma FW from F-Droid to block internet connections.

I'm aware that I could easily use a firewall app to manually revoke the app's Internet access, but if that's the step that I have to take, then at that point, I may as well go one tiny step further and use a closed-source music player. Like, if my firewall app ensures that nothing is getting off my device anyway, then I may as well use a more feature-rich closed-source version.

It's simply a matter of absolute principle. Trust (FOSS) but verify (100% offline).

> BTW, software can require internet connection and still be FOSS while respecting user privacy.

Yes, but it should be kept to a bare minimum, which in the F-Droid version of Symphony's case, is zero.

There's little reason to self-penalize the app's F-Droid status when the Internet permission isn't necessary for the F-Droid releases.

> I don't know what to say, if this little thing is too much of a deal breaker, download another app from the pleathora that exist.

The problem is there are no real alternatives.

On F-Droid, the music players that are 100% offline are either stuck in outdated Holo-era designs, or if they are modern and slick, then they are either too barebones with not enough features (Auxio) or self-declared dead by the developer with no updates to come (Metro).

Symphony is the only one that satisfies the conditions that I'm looking for (modern design, feature-rich, actively maintained), except for one: being 100% offline.

Trust me when I say that if I knew of any such music player, I would have saved my time and just used that, instead of writing that long, long main post up there. :)

**Comment by zyrouge** [comment]
Might reconsider this, but I'd hate to not have reproducible builds.

**Comment by puelp** [comment]
Thank you!

**Comment by zyrouge** [comment]
> > but I'd hate to not have reproducible builds
> 
> If you already have RB, then why not make F-Droid the default update source for your app? Self-updating is a mess of Internet, Install, Storage, etc. permissions. F-Droid updates need none of those.

F-droid updates are slow. That's why a message that there is an update is implemented instead of self-updates.

**Comment by zyrouge** [comment]
> Oh, if it's just a message, then I'm even more confused. F-Droid (or Neo Store) does provide a notification when there is an update available.

Thing is I like using GitHub Releases more than F-droid. I don't want to let my the users who downloaded from GitHub Releases to stay behind the latest. Neither do I want to prevent them from moving to F-droid version. That's why I don't want to have different flavours since that would make reproducible builds not possible and will also have different signatures.

**Comment by zyrouge** [comment]
> There's nothing in the app that's mission critical, though, right? It's not like a browser that needs to stay on top of the latest CVE. Those users who want more frequent updates (I wonder what %age of the userbase they make up), can equally tap a link to GitHub Releases in your app, which doesn't need Internet permission since it merely passes an Intent (I think).

Nobody likes to occasionally check a link to check if there is an update. And I'm sure that this is not going to be removed. The worst is having a different build in F-droid. The recommended places will remain to be GitHub or IzzyOnDroid.

**Comment by pabloscloud** [comment]
So basically internet access is required to check for updates?

**Comment by zyrouge** [comment]
> So basically internet access is required to check for updates?

Yes.

**Comment by pabloscloud** [comment]
Well, if there is a reason that this app needs internet access it's no even up to debate to remove the internet access, is it?

Yeah, if one uses a store to automatically update, this function isn't needed, but the work that would need to be done every update to archive this is hilarious imho.

Not that I don't like the idea of having the permission removed, I totally do, but it's not worth the effort.

And as mentioned before, I just use a Firewall to block the connections.

Only thing I could imagine is a little popup or sth saying that this app checks for updates on its own and maybe a toggle to disable that function. This wouldn't remove the access, but at least the user has the choice if data is send and using f-droid one can also verify that no other data is send by looking at the source code.

**Comment by zyrouge** [comment]
There is already a setting to disable update notification.

**Comment by pabloscloud** [comment]
does this also disable the search for new versions or just basically doesn't show a notification?

If this also disables the lookup for new versions, I don't see an issue here. But who am I to say that

**Comment by zyrouge** [comment]
> does this also disable the search for new versions or just basically doesn't show a notification?
> 
> If this also disables the lookup for new versions, I don't see an issue here. But who am I to say that

It disables the notification only since the settings page will show if the current is not up-to-date which is handy considering how many users never know when a version is released.

**Comment by pabloscloud** [comment]
Well, that can be turned off too cause if one uses f-droid, the updating process isn't handled by the app. This app doesn't need to be up-to-date all the time either way.

**Comment by zyrouge** [comment]
> Well, that can be turned off too cause if one uses f-droid, the updating process isn't handled by the app.

That's there for the sake of just being handy for someone.

> This app doesn't need to be up-to-date all the time either way.

Not quite true considering how often bugs are found and new things are added constantly.

**Comment by pabloscloud** [comment]
Yes, from a security perspective.

**Comment by IzzySoft** [comment]
Maybe I can offer some mitigation here. With your app now on F-Droid, I originally came here to ask whether it's OK to remove it from my repo. Instead, reading your concerns, I just re-decided:

In-App-Updaters, especially when enabled by default, violate [F-Droid's inclusion criteria](https://f-droid.org/docs/Inclusion_Policy/?title=Inclusion_Policy) – which say the app itself must not load binaries, quote:

> The software must not download additional executable binary files (e.g. addons, auto-updates, etc.) without explicit user consent. Consent means it needs to be opt-in (it must not be harder to decline than to accept or presented in a way users are likely to press accept without reading) and structured in a way that clearly explains to users that they’re choosing to bypass F-Droid’s checks if they activate it.

With your app having achieved reproducible builds, keeping it in my repo would render the integrated updater obsolete I'd say. Your app is listed with my repo since 2022-11-16, so you might have noticed that updates become available there within less than 24h of your tagging a new release and attaching the APK to it. Would you consider that sufficient and thus drop the integrated updater (or at least make it opt-in)? Then I'd be willing to keep your app in my repo despite it's already available at F-Droid.org.

As a remark: my repo isn't enabled by default in the (official) F-Droid client, so enabling it counts as opt-in. Thanks to reproducible builds, such cross-updates become possible.

So will you accept my "deal", @zyrouge?

**Comment by zyrouge** [comment]
> Maybe I can offer some mitigation here. With your app now on F-Droid, I originally came here to ask whether it's OK to remove it from my repo. Instead, reading your concerns, I just re-decided:
> 
> In-App-Updaters, especially when enabled by default, violate [F-Droid's inclusion criteria](https://f-droid.org/docs/Inclusion_Policy/?title=Inclusion_Policy) – which say the app itself must not load binaries, quote:
> 
> > The software must not download additional executable binary files (e.g. addons, auto-updates, etc.) without explicit user consent. Consent means it needs to be opt-in (it must not be harder to decline than to accept or presented in a way users are likely to press accept without reading) and structured in a way that clearly explains to users that they’re choosing to bypass F-Droid’s checks if they activate it.
> 
> With your app having achieved reproducible builds, keeping it in my repo would render the integrated updater obsolete I'd say. Your app is listed with my repo since 2022-11-16, so you might have noticed that updates become available there within less than 24h of your tagging a new release and attaching the APK to it. Would you consider that sufficient and thus drop the integrated updater (or at least make it opt-in)? Then I'd be willing to keep your app in my repo despite it's already available at F-Droid.org.
> 
> As a remark: my repo isn't enabled by default in the (official) F-Droid client, so enabling it counts as opt-in. Thanks to reproducible builds, such cross-updates become possible.
> 
> So will you accept my "deal", @zyrouge?

That would still leave the users who downloaded the app from GitHub down. It's not possible to keep it opted-in for GitHub version and opted-out for others without having seperate builds. This is my only concern. I still consider GitHub and IzzyOnDroid as recommended sources due to faster updates. It's okay with me if you are willing to remove it from IzzyOnDroid.

**Comment by pabloscloud** [comment]
But you can just implement a popup where the user decides to enable or disable auto-updates.

**Comment by IzzySoft** [comment]
> That would still leave the users who downloaded the app from GitHub down.

It would leave nobody down if it's strictly opt-in. Whoever feels down can opt-up :stuck_out_tongue_winking_eye: And as Pablo points out: Make it a boolean field, initiates as NULL. If it's NULL when the app is started, have a (modal) popup to ask the user's decision. Otherwise honor the decision – plus give the user the possibility to revise it later (option in settings).

> It's okay with me if you are willing to remove it from IzzyOnDroid.

It would be due to removal today then (I usually give apps an overlap of 10..14 days). So if you insist on your updater, I see no reason to keep it in my repo: it's on F-Droid.org now, ensuring a FOSS build with all the F-Droid checks and reproducible at that (otherwise your updater would be useless there due to signature mismatch) – and for those who prefer Github releases and don't want F-Droid, you have your own updater.

**Comment by zyrouge** [comment]
> But you can just implement a popup where the user decides to enable or disable auto-updates.

I think this might be the best way to solve this.

**Comment by zyrouge** [comment]
> 

I don't really want it be removed from IzzyOnDroid, the above solution would be implemented within next release. And fyi, the updater can already be disabled in settings.

**Comment by pabloscloud** [comment]
> And fyi, the updater can already be disabled in settings.

Yes, but:
> It disables the notification only since the settings page will show if the current is not up-to-date

**Comment by IzzySoft** [comment]
@zyrouge so is it opt-in now? Then I'll mark it to be kept. Now's the time to decide, it's been a month that your app reached F-Droid – and I wonder if that shouldn't mean opt-in must have been established already (as per the rules there).

**Comment by zyrouge** [comment]
> @zyrouge so is it opt-in now? Then I'll mark it to be kept. Now's the time to decide, it's been a month that your app reached F-Droid – and I wonder if that shouldn't mean opt-in must have been established already (as per the rules there).

No updates has been released yet. I'm currently busy with my real life stuff. This is planned for next update which will be released within next week (hopefully).

**Comment by IzzySoft** [comment]
Thanks! Can you give me a ping then, please?

**Comment by zyrouge** [comment]
> Thanks! Can you give me a ping then, please?

Sure.

**Comment by zyrouge** [comment]
Ability to opt-in or opt-out (opted-out by default) for version checking has been implemented as of v2023.5.89.

/cc @IzzySoft

**Comment by zyrouge** [comment]
To make this issue clearer,

- There is no self-updater implemented within the app.
- The app checks for the latest version from Github Releases at the startup and displays a toast.

If the F-droid version needs to completely free from this permission, separate flavor of the app must be build exclusive for the it. This would prevent any users to migrate from Github version to F-droid or the vice-versa. Also the fact that F-droid updates take at least 7-10 days to publish the version.

The primary release source of the app will stay as Github Releases (and IzzyOnDroid), with F-droid as secondary source. And if the app is released on Play Store, it would be using the same flavor as the primary release source.

If F-droid users still want a separate build flavor, please let me know your thoughts (if you really want or if you are okay current state) by posting a comment. This would be implemented only if there is enough demand from the users.

**Comment by Mederim** [comment]
I would like to opinion in favor to full offline music player and also said that is no need worry about update checks since we have a simple solution for that problem what is a RSS feed like said in [here](https://www.privacyguides.org/en/android/#manually-with-rss-notifications) and also is good for havin just one build system because not all users wanna use the signature system from f-droid because [secure and privacy problems](https://www.youtube.com/watch?v=lAbgeJau3eE) and just use github releases

**Comment by zyrouge** [comment]
> I would like to opinion in favor to full offline music player and also said that is no need worry about update checks since we have a simple solution for that problem what is a RSS feed like said in [here](https://www.privacyguides.org/en/android/#manually-with-rss-notifications) and also is good for havin just one build system because not all users wanna use the signature system from f-droid because [secure and privacy problems](https://www.youtube.com/watch?v=lAbgeJau3eE) and just use github releases

Just to make it clear, the check for updates option will not be removed from primary flavor of the app. Removing it would lead separation between official users and F-droid users. Thus, F-droid flavor would possibly be fully offline and will be using *F-droid's signature*.

**Comment by IzzySoft** [comment]
> Ability to opt-in or opt-out (opted-out by default)

That should meet the F-Droid criteria towards this, thanks!

> Also the fact that F-droid updates take at least 7-10 days to publish the version.

Ugh? When did you check that last? Currently it spits out updates almost daily. So unless the build fails, it's rather 2-3 days.

> This would prevent any users to migrate from Github version to F-droid or the vice-versa.

Not necessarily. If you'd use the same `applicationId` and [reproducible builds](https://gitlab.com/-/snippets/2544837), cross-updates would still be possible. But as I wrote: if it's strictly opt-in and not even an updater but just an "update informer", this does not conflict with the F-Droid inclusion criteria.

**Comment by zyrouge** [comment]
> Ugh? When did you check that last? Currently it spits out updates almost daily. So unless the build fails, it's rather 2-3 days.

My bad. Last update took nearly 7 days to get published. Just checked and it seems like they update every 2-3 days.

> Not necessarily. If you'd use the same applicationId and reproducible builds, cross-updates would still be possible. But as I wrote: if it's strictly opt-in and not even an updater but just an "update informer", this does not conflict with the F-Droid inclusion criteria.

Changing permissions essentially means different manifests with different ids. This would prevent reproducible builds and cross updates since that flavor would only be on F-droid and nowhere else, which is something I really hate to do.

**Comment by IzzySoft** [comment]
> My bad. Last update took nearly 7 days to get published.

Up to a few months ago, that was indeed the case. It luckily improved, the team always gives its best (though it cannot do everything at once :see_no_evil:).

> Changing permissions essentially means different manifests with different ids. This would prevent reproducible builds and cross updates

You miss a point here, so let me outline:

* you have a `fdroid` flavor which is reproducible. So that APK gets distributed by F-Droid **with your signature**.
* you have some other flavor, which uses the same `applicationId`. You sign that APK with the same key.
* I have this "other flavor" installed. As it has the same signer/signature, F-Droid offers me the next update – which of course is the `fdroid` flavor

So it would work. But that said, it won't be necessary for the issue on hand, as the opt-in update notifications as outlined above are compliant with F-Droid's inclusion criteria.

**Comment by zyrouge** [comment]
Yup, I did understand that and that's why I said that the flavor will only be in F-droid (meaning it will not be published in Github Releases) preventing reproducible builds.

**Comment by IzzySoft** [comment]
Ah, if there are no reproducible builds, it of course won't work, right. Though I don't get why you'd want to do that. Luckily we don't need to discuss this, as we all are happy with the one flavor, as outlined above :smiley:

**Comment by Mederim** [comment]
I don't understand your point when you said that about separation between fdroid an official users? Not removing the updated check and go with not fully offline music player will lead into separation because people will use f-droid version instead the official version because we will have two versions of the same app and not the same app for everyone. Also i don't understand well why you wanna keep check updates since is no need for most people to use, like i said using a RSS feed or [Obtainium](https://github.com/ImranR98/Obtainium) 

> Just to make it clear, the check for updates option will not be removed from primary flavor of the app. Removing it would lead separation between official users and F-droid users.

**Comment by zyrouge** [comment]
> I don't understand your point when you said that about separation between fdroid an official users? Not removing the updated check and go with not fully offline music player will lead into separation because people will use f-droid version instead the official version because we will have two versions of the same app and not the same app for everyone. Also i don't understand well why you wanna keep check updates since is no need for most people to use, like i said using a RSS feed or [Obtainium](https://github.com/ImranR98/Obtainium)
> 
> > Just to make it clear, the check for updates option will not be removed from primary flavor of the app. Removing it would lead separation between official users and F-droid users.

Not everyone have a habit of using RSS feeds. I'm pretty sure 90% of the users use the notification message to update the app. Not everyone is a power user.

**Comment by Mederim** [comment]
> Not everyone have a habit of using RSS feeds. I'm pretty sure 90% of the users use the notification message to update the app. Not everyone is a power user.

I think that linda of users don't concern about f-droid signed apks so that user will use f-droid for updates anyway so the majority users of git release will be people who don't wanna use f-droid signed apk who might use RSS feeds for updates. I really don't see the point of enable check updates for just github since the normal user can use f-droid for update, and will no need the github version, i assume that the generic user don't wanna waste time and just give f-droid the responsability for all updates. I assume that mostly github releases users don't wanna user f-droid signed and are using some kinda of solution outside f-droid for updates since why will that user be using github release without a really good reason to denny f-droid?

**Comment by Mederim** [comment]
Also you could just simple recommend or put a link for using RSS feed and obtainium in the readme, it will educate people who don't know this possibility

**Comment by zyrouge** [comment]
> > Not everyone have a habit of using RSS feeds. I'm pretty sure 90% of the users use the notification message to update the app. Not everyone is a power user.
> 
> I think that kinda of users don't concern about f-droid signed apks so that user will use f-droid for updates anyway so the majority users of git release will be people who don't wanna use f-droid signed apk who might use RSS feeds for updates. I really don't see the point of enable check updates for just github since the normal user can use f-droid for update, and will no need the github version, i assume that the generic user don't wanna waste time and just give f-droid the responsability for all updates. I assume that mostly github releases users don't wanna user f-droid signed and are using some kinda of solution outside f-droid for updates since why will that user be using github release without a really good reason to denny f-droid?

Good question. The main problem I had with F-droid was how slow the updates were. It was around 7 days. However since they have reduced to around 3 days now, I could possibly give preference to it. This still doesn't answer the question of "how to keep users up-to-date with the latest version of the app?". Statistically speaking, over 1k users have downloaded the previous version from Github Releases. And most users use the update message to know that an update has been published. Most (including me) never use RSS feeds or even check Play Store for updating an app though I download apps from F-droid, I don't have F-droid installed on my phone. This is why this has been implemented at the first place. Unless a viable solution is implemented, this is not getting removed, at least from the primary flavor.

**Comment by Mederim** [comment]
I know that is difficult to have a solution to everyone but i still will be prefer disable the update check for every release than focus on the normal user get it update, like i saíd the normal user will still be using and have f-droid for solution this and will no concern about it. I give you the example of [Metro](https://github.com/MuntashirAkon/Metro) who is only for f-droid users and is used for a bunch of people and also you van see this [request in metro repo](https://github.com/MuntashirAkon/Metro/pull/164) and ser that people who wanna git releases recommend obtainium for updates who is a reliable way. In mostly apps that don't have git releases everytime i ser that is a poweruser need it than a normal user.

**Comment by IzzySoft** [comment]
@Mederim I don't know which issue you're seeing. If I understood correctly, that update check will now be opt-in, i.e. disabled by default. So you just don't need to enable it. And those few users who want it, can simply switch it on. With that, all sides should be happy, no? Or did I miss something?

**Comment by zyrouge** [comment]
A [PR](https://gitlab.com/fdroid/fdroiddata/-/merge_requests/13164) has been opened to remove anti-feature tag from the F-droid page, but it seems like no PRs has been merged for the past 4 weeks.

**Comment by IzzySoft** [comment]
> it seems like no PRs has been merged for the past 4 weeks.

The [first page or merged MRs](https://gitlab.com/fdroid/fdroiddata/-/merge_requests?scope=all&state=merged) does not even reach back 4 weeks, so there were plenty of MRs merged :wink: Thing is, MRs have to be reviewed first. We are few, and the queue is quite long…

So as I got the ping here, and checking this one doesn't take longer than writing: I've pulled the latest version and checked, there's the described toggle and it's turned off by default, so I hereby confirm the AF became obsolete (didn't see any other network connections being opened either). So I will merge that now, thanks!
